scripts/4.0-migration.sh

```diff
@@ -0,0 +1,338 @@
+cyral_repository and cyral_repository_binding resource that
+will be migrated. Additionally, cyral_sidecar listeners
+will be created, which are now required to bind sidecars to
+repositories."
+echo
+echo -e "${green}Please set CYRAL_TF_FILE_PATH equal to the file path of your .tf file.${clear}"
+echo
+read -p "Are you ready to continue? [N/y] " -n 1 -r
+echo
+if [[ ! $REPLY =~ ^[Yy]$ ]]
+then
+    echo "Exiting..."
+    [[ "$0" = "$BASH_SOURCE" ]] && exit 1 || return 1 # handle exits from shell or function but don't exit interactive shell
+fi
+echo
+echo "Searching for cyral_repository and cyral_repository_binding resources to migrate..."
+echo
+
+terraform state pull > terraform.tfstate.cyral.migration.backup
+cp ${CYRAL_TF_FILE_PATH} cyral_terraform_migration_backup_configuration.txt
+
+repo_import_args=()
+binding_import_args=()
+listener_import_args=()
+access_gateway_import_args=()
+
+repo_resource_defs=()
+binding_resource_defs=()
+listener_resource_defs=()
+access_gateway_resource_defs=()
+
+repo_resource_names=()
+binding_resource_names=()
+listener_resource_names=()
+access_gateway_resource_names=()
+
+repos_to_delete=()
+bindings_to_delete=()
+
+# Create array of repo and binding resource to migrate
+resources_to_migrate=($(terraform state list | grep "cyral_repository\|cyral_repository_binding"))
+# Store terraform state JSON representation
+tf_state_json=$(terraform show -json | jq ".values.root_module.resources[]")
+
+# Find all cyral_repository and cyral_repository_local_accounts
+for resource_address in ${resources_to_migrate[@]}; do
+  if [[ $resource_address == cyral_repository.* ]]
+  then
+    # We will need to delete this repo from the .tf file, so we
+    # store the full resource address.
+    repos_to_delete+=($resource_address)
+    # Get repo ID. This will be used to import the updated repo.
+    repo_id=($(jq -r "select(.address == \"$resource_address\") | .values.id"<<<$tf_state_json))
+    # Remove [] from the resource address and substitute [ for _
+    # E.g. cyral_repository.repo[0] -> cyral_repository.repo_0
+    resource_address=$(sed -e 's/[]]//g;s/[[]/_/g'<<<$resource_address)
+    # Save an empty resource definition for a new repo, so that
+    # it can be added to the .tf file.
+    repo_resource_defs+=("resource \"cyral_repository\" \"${resource_address##"cyral_repository."}\" {}")
+    # Store import resource address and repo ID, which will be the argument to terraform import.
+    repo_import_args+=("${resource_address} ${repo_id}")
+    repo_resource_names+=("${resource_address}")
+  elif [[ $resource_address == cyral_repository_binding.* ]]
+  then
+    # We will need to delete this sidecar binding from the .tf file, store its name
+    bindings_to_delete+=($resource_address)
+    # Get ids required to import binding resource.
+    id_values_arr=($(jq -r "select(.address == \"$resource_address\") | .values.sidecar_id, .values.repository_id, .values.sidecar_as_idp_access_gateway"<<<$tf_state_json))
```

Comment:
Will the IDs for the migrated bindings always be equal to the `repository_id`? If they're not, the import will not work on 4.0 - neither the access gateway import -, since it expects the binding resource ID to be in the format `{sidecar_id}/{binding_id}`, and in this case we are using `{sidecar_id}/{repository_id}` to import the bindings, so the `binding_id` must be equal to the `repository_id` for this to work.

Reply:
I just confirmed that the bindings and the new access gateway resources will be migrated with the same ID as the repo. The `4.0-migration.sh` script worked just fine.

Reply:
Yes, that is correct! Confirmed in todays stand up :)